### PR TITLE
Add extensions support for patch

### PR DESCRIPTION
**Pull request by shaosss**
Add extensions support for Patch as discussed [here](https://chat.fhir.org/#narrow/stream/179166-implementers/topic/Use.20Patch.20with.20extensions). 

I tested with some scenarios which could cause troubles. As long as FHIR Model types are consistent in annotation and dataTypes - there should be no issues.

I think @kennethmyhra is aware of the discussion mentioned at top.

### CODE REVIEW

File: src/Spark.Engine.Test/Service/PatchServiceTests.cs
Comment (kennethmyhra) [inline src/Spark.Engine.Test/Service/PatchServiceTests.cs:191]: Even though they all go down the same code path (`CreateValueExpression`), have you tested `insert` and `replace` as well with FHIR extensions?

File: src/Spark.Engine.Test/Service/PatchServiceTests.cs
Comment (kennethmyhra) [inline src/Spark.Engine.Test/Service/PatchServiceTests.cs]: I'm a bit unfamiliar with the Patch code and the extension methods, but I guess there isn't an easier way to do this?

**Review by kennethmyhra — COMMENTED** [review]
Thanks, @shaosss!

Have we tested if we can apply extensions on attributes below root level of the resource? I.e. adding extensions to Specimen.process.description

File: src/Spark.Engine.Test/Service/PatchServiceTests.cs
Comment (shaosss) [reply src/Spark.Engine.Test/Service/PatchServiceTests.cs]: New extension methods could be added, so it would allow to add generic parts as a value. Like AddAddPatchParameter(string path, string name, List<Parameters.ParameterComponent> parts).
Though those are part of the Hl7.Fhir.Rest, so it would need to go through there first.
There are some methods similar to what is needed, but they are all private. :(

File: src/Spark.Engine.Test/Service/PatchServiceTests.cs
Comment (shaosss) [reply src/Spark.Engine.Test/Service/PatchServiceTests.cs:191]: Good idea. I did not, but I will check with unit tests.

**Comment by shaosss** [comment]
> 
> 
> Thanks, @shaosss!
> 
> Have we tested if we can apply extensions on attributes below root level of the resource? I.e. adding extensions to Specimen.process.description

According to the specification and discussion - yes. Before this PR it would fail with exception though.

File: src/Spark.Engine.Test/Service/PatchServiceTests.cs
Comment (shaosss) [reply src/Spark.Engine.Test/Service/PatchServiceTests.cs:191]: Checked and added unit tests. Found one potential issue and also fixed it.
Regarding insert. I am not sure current code handles it according to spec. We seems to be getting indexes from path, while specification says that the index should be taken from index part. (https://www.hl7.org/fhir/fhirpatch.html). I did add the handling of this case for the insert, without removing the current logic, but if it is not according to spec it might be better to remove it, IMHO. I also marked it with ToDo in unit tests. (maybe I should have done it in the code.) The examples for the insert could be found at FHIR site: (https://www.hl7.org/fhir/test-cases.zip). I looked at patch\fhir-path-tests.xml.

Comment (kennethmyhra) [reply src/Spark.Engine.Test/Service/PatchServiceTests.cs:191]: Agree, that should not be correct to how the current specification is to be interpreted. If you can remove it I would appreciate it, better to remove it now before people start relying heavily on it.

**Review by kennethmyhra — APPROVED** [review]
LGTM, let me know if you have time for that last TODO, unless I think this is ready to be merged

**Comment by shaosss** [comment]
@kennethmyhra I accidently closed PR, sorry. :) I fixed TODO, so it is ready to merge. I think it does require your re-approval or completion

**Comment by kennethmyhra** [comment]
Thanks, @shaosss!

**Pull request merged**
